File: desktop/Desktop/Sources/LiveNotes/LiveNotesMonitor.swift
import Foundation
import Combine
import GRDB

/// Dedicated monitor for live notes generation during recording sessions.
/// Accumulates transcript words and triggers AI note generation at word thresholds.
/// Only views that explicitly observe this class will update when notes change.
@MainActor
class LiveNotesMonitor: ObservableObject {
    static let shared = LiveNotesMonitor()

    /// Live notes for real-time display
    @Published private(set) var notes: [LiveNote] = []

    /// Whether AI note generation is enabled
    @Published var isAiEnabled: Bool = true

    /// Whether a note is currently being generated
    @Published private(set) var isGenerating: Bool = false

    /// Current recording session ID
    private var currentSessionId: Int64?

    /// Word buffer for tracking transcript content
    private var wordBuffer: [String] = []

    /// Segment order tracking for note context
    private var lastProcessedSegmentOrder: Int = -1

    /// Current segment order being processed
    private var currentSegmentOrder: Int = 0

    /// End time of the last segment we extracted words from (cursor for incremental processing)
    private var lastProcessedSegmentEnd: Double?

    /// Minimum words before triggering AI generation
    private let wordThreshold = 50

    /// Max words to keep in buffer (older words are trimmed)
    private let maxWordBufferSize = 500

    /// Max existing notes to keep for context (oldest trimmed)
    private let maxExistingNotesContext = 20

    /// Existing notes for context (to avoid repetition)
    private var existingNotesContext: [String] = []

    /// GeminiClient for AI generation (lazily initialized)
    private var geminiClient: GeminiClient?

    /// Cancellables for subscriptions
    private var cancellables = Set<AnyCancellable>()

    /// AI prompt for note generation (from m13v/meeting)
    private let noteGenerationPrompt = """
        generate a single, concise note about what happened in this segment.
        be factual and specific.
        focus on the key point or action item.
        keep it a few word sentence.
        do not use quotes.
        do not use wrapping words like "discussion on", jump straight into note.
        avoid repeating information from existing notes.
        """

    private init() {
        // Subscribe to transcript changes
        LiveTranscriptMonitor.shared.$segments
            .receive(on: DispatchQueue.main)
            .sink { [weak self] segments in
                self?.handleSegmentsUpdate(segments)
            }
            .store(in: &cancellables)
    }

    // MARK: - Session Lifecycle

    /// Start a new notes session
    func startSession(sessionId: Int64) {
        log("LiveNotesMonitor: Starting session \(sessionId)")
        currentSessionId = sessionId
        notes = []
        wordBuffer = []
        lastProcessedSegmentOrder = -1
        currentSegmentOrder = 0
        lastProcessedSegmentEnd = nil
        existingNotesContext = []

        // Initialize Gemini client if not already done
        if geminiClient == nil {
            do {
                // Use Gemini 3 Pro for better note generation quality
                geminiClient = try GeminiClient(model: "gemini-pro-latest")
                log("LiveNotesMonitor: GeminiClient initialized with gemini-pro-latest")
            } catch {
                logError("LiveNotesMonitor: Failed to initialize GeminiClient", error: error)
            }
        }

        // Load any existing notes from DB (for crash recovery)
        Task {
            await loadExistingNotes()
        }
    }

    /// End the current notes session
    func endSession() {
        log("LiveNotesMonitor: Ending session \(currentSessionId ?? -1) with \(notes.count) notes")
        currentSessionId = nil
        wordBuffer = []
        lastProcessedSegmentOrder = -1
        currentSegmentOrder = 0
        lastProcessedSegmentEnd = nil
    }

    /// Clear all notes (used when recording stops)
    func clear() {
        notes = []
        wordBuffer = []
        existingNotesContext = []
        lastProcessedSegmentOrder = -1
        currentSegmentOrder = 0
        lastProcessedSegmentEnd = nil
    }

    // MARK: - Note Operations

    /// Add a manual note
    func addManualNote(text: String) {
        guard let sessionId = currentSessionId else {
            log("LiveNotesMonitor: Cannot add note - no active session")
            return
        }

        Task {
            do {
                let record = try await NoteStorage.shared.createNote(
                    sessionId: sessionId,
                    text: text,
                    isAiGenerated: false,
                    segmentStartOrder: currentSegmentOrder
                )

                if let note = record.toLiveNote() {
                    await MainActor.run {
                        self.notes.append(note)
                        self.existingNotesContext.append(text)
                        // Trim context to prevent unbounded growth
                        if self.existingNotesContext.count > self.maxExistingNotesContext {
                            self.existingNotesContext.removeFirst(self.existingNotesContext.count - self.maxExistingNotesContext)
                        }
                    }
                }
            } catch {
                logError("LiveNotesMonitor: Failed to add manual note", error: error)
            }
        }
    }

    /// Update an existing note
    func updateNote(id: Int64, text: String) {
        Task {
            do {
                try await NoteStorage.shared.updateNote(id: id, text: text)

                await MainActor.run {
                    if let index = self.notes.firstIndex(where: { $0.id == id }) {
                        var updatedNote = self.notes[index]
                        updatedNote.text = text
                        updatedNote.updatedAt = Date()
                        self.notes[index] = updatedNote
                    }
                }
            } catch {
                logError("LiveNotesMonitor: Failed to update note", error: error)
            }
        }
    }

    /// Delete a note
    func deleteNote(id: Int64) {
        Task {
            do {
                try await NoteStorage.shared.deleteNote(id: id)

                await MainActor.run {
                    self.notes.removeAll { $0.id == id }
                }
            } catch {
                logError("LiveNotesMonitor: Failed to delete note", error: error)
            }
        }
    }

    // MARK: - Private Methods

    /// Load existing notes from DB (for crash recovery)
    private func loadExistingNotes() async {
        guard let sessionId = currentSessionId else { return }

        do {
            let existingNotes = try await NoteStorage.shared.getLiveNotes(sessionId: sessionId)
            await MainActor.run {
                self.notes = existingNotes
                self.existingNotesContext = existingNotes.map { $0.text }
            }
            log("LiveNotesMonitor: Loaded \(existingNotes.count) existing notes from DB")
        } catch {
            logError("LiveNotesMonitor: Failed to load existing notes", error: error)
        }
    }

    /// Handle transcript segments update
    private func handleSegmentsUpdate(_ segments: [SpeakerSegment]) {
        guard currentSessionId != nil, isAiEnabled else { return }

        // Track segment order
        currentSegmentOrder = segments.count

        // Only extract words from segments we haven't processed yet.
        // Use the last processed segment's end time as a cursor to find new content.
        let newSegments: ArraySlice<SpeakerSegment>
        if let lastEnd = lastProcessedSegmentEnd {
            // Find segments that are new or were updated (end time > last processed)
            if let startIdx = segments.firstIndex(where: { $0.end > lastEnd }) {
                newSegments = segments[startIdx...]
            } else {
                return  // No new segments
            }
        } else {
            newSegments = segments[...]
        }

        let newWords = newSegments.flatMap { $0.text.split(separator: " ").map(String.init) }
        guard !newWords.isEmpty else { return }

        // Update cursor to the end of the last segment we processed
        if let lastSeg = segments.last {
            lastProcessedSegmentEnd = lastSeg.end
        }

        wordBuffer.append(contentsOf: newWords)

        // Trim word buffer to prevent unbounded growth (keep most recent words)
        if wordBuffer.count > maxWordBufferSize {
            wordBuffer.removeFirst(wordBuffer.count - maxWordBufferSize)
        }

        // Check if we have enough words to generate a note
        let wordsSinceLastNote = wordBuffer.count - (lastProcessedSegmentOrder >= 0 ? lastProcessedSegmentOrder : 0)
        if wordsSinceLastNote >= wordThreshold && !isGenerating {
            generateNote(from: segments)
        }
    }

    /// Generate an AI note from recent transcript
    private func generateNote(from segments: [SpeakerSegment]) {
        guard let sessionId = currentSessionId,
              let client = geminiClient,
              !isGenerating else { return }

        isGenerating = true

        // Get recent transcript text (last ~50 words)
        let recentText = wordBuffer.suffix(wordThreshold).joined(separator: " ")
        let segmentStartOrder = max(0, currentSegmentOrder - 3)
        let segmentEndOrder = currentSegmentOrder

        // Build context from existing notes
        let existingNotesText = existingNotesContext.isEmpty
            ? "No existing notes yet."
            : "Existing notes:\n" + existingNotesContext.map { "- \($0)" }.joined(separator: "\n")

        let prompt = """
            Transcript segment:
            \(recentText)

            \(existingNotesText)

            \(noteGenerationPrompt)
            """

        Task {
            do {
                let response = try await client.sendTextRequest(
                    prompt: prompt,
                    systemPrompt: "You are a concise note-taker. Generate a single short note (3-10 words) about the key point in the transcript. Do not use quotes. Be direct and specific."
                )

                // Clean up the response
                let noteText = response
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                    .replacingOccurrences(of: "\"", with: "")
                    .replacingOccurrences(of: "'", with: "")

                guard !noteText.isEmpty else {
                    await MainActor.run { self.isGenerating = false }
                    return
                }

                // Save to DB
                let record = try await NoteStorage.shared.createNote(
                    sessionId: sessionId,
                    text: noteText,
                    isAiGenerated: true,
                    segmentStartOrder: segmentStartOrder,
                    segmentEndOrder: segmentEndOrder
                )

                if let note = record.toLiveNote() {
                    await MainActor.run {
                        self.notes.append(note)
                        self.existingNotesContext.append(noteText)
                        // Trim context to prevent unbounded growth (keep most recent notes)
                        if self.existingNotesContext.count > self.maxExistingNotesContext {
                            self.existingNotesContext.removeFirst(self.existingNotesContext.count - self.maxExistingNotesContext)
                        }
                        self.lastProcessedSegmentOrder = self.wordBuffer.count
                        self.isGenerating = false
                    }
                } else {
                    await MainActor.run { self.isGenerating = false }
                }
            } catch let dbError as DatabaseError where dbError.resultCode == .SQLITE_CONSTRAINT {
                // Session was deleted during async AI generation — not an error
                log("LiveNotesMonitor: Session \(sessionId) deleted during note generation, skipping")
                await MainActor.run { self.isGenerating = false }
            } catch {
                logError("LiveNotesMonitor: Failed to generate note", error: error)
                await MainActor.run { self.isGenerating = false }
            }
        }
    }

    // MARK: - Computed Properties

    /// Check if there are any notes
    var isEmpty: Bool {
        notes.isEmpty
    }

    /// Get the latest note
    var latestNote: LiveNote? {
        notes.last
    }

    /// Get notes for current session
    func getNotesForCurrentSession() -> [LiveNote] {
        return notes
    }

    // MARK: - Diagnostics

    /// Word buffer size (for memory diagnostics)
    var wordBufferCount: Int { wordBuffer.count }

    /// Existing notes context size (for memory diagnostics)
    var existingNotesContextCount: Int { existingNotesContext.count }
}

File: desktop/windows/src/renderer/src/lib/liveNotes/liveNotesMonitor.ts
// PR8 LiveNotes — the monitor: subscribes to the live transcript store and drives
// AI note generation, plus user-typed note add/edit/delete. A faithful port of the
// macOS LiveNotesMonitor (Desktop/Sources/LiveNotes/LiveNotesMonitor.swift), a
// subscribable singleton so the notes panel can `subscribe()` exactly like it
// reads `liveConversation` — reusing the ALREADY-mirrored transcript stream (fed
// by LiveMirrorHost ← IPC ← the capture window). No new WebSocket/IPC for reads.
//
// Where it runs: mounted once at the app-shell root (LiveNotesHost), so generation
// runs whenever the main window is open regardless of whether the notes PANEL is
// open — matching Mac's "generation is tied to the transcription lifecycle, not
// the panel". (It is main-window-scoped, like the transcript mirror it reads;
// headless continuous recording with the window closed pauses generation, same as
// the mirror. Documented divergence — see the report.)
//
// Session model: the capture window owns the real client_conversation_id, which is
// NOT carried in the mirrored store, so the monitor mints its OWN session id when a
// new transcript starts and persists a transcription_sessions row (the anchor for
// the notes' cascading FK). Nothing joins sessions to conversations, so a
// monitor-minted id is self-consistent.

import type { LiveNote, TranscriptLine } from '../../../../shared/types'
import { liveConversation, type LiveStatus } from '../liveConversation'
import { generate as geminiGenerate } from '../geminiClient'
import { trackEvent } from '../analytics'
import { LiveNotesAccumulator, type LiveNotesGenerationRequest } from './liveNotesAccumulator'

// Mac's prompts, verbatim (LiveNotesMonitor.swift:64-72, 262).
const NOTE_GENERATION_PROMPT = `generate a single, concise note about what happened in this segment.
be factual and specific.
focus on the key point or action item.
keep it a few word sentence.
do not use quotes.
do not use wrapping words like "discussion on", jump straight into note.
avoid repeating information from existing notes.`

const SYSTEM_PROMPT =
  'You are a concise note-taker. Generate a single short note (3-10 words) about the key point in the transcript. Do not use quotes. Be direct and specific.'

// Gemini Flash on the premium tier (Mac's ModelQoS.Gemini.proactive default);
// thinkingBudget 0 = thinking disabled (cheapest), matching Mac's LiveNotes which
// passes no custom budget. Max-tier pro upgrade is deferred (documented).
const NOTE_MODEL = 'gemini-2.5-flash'

/** Turns a transcript segment + existing notes into a note string. Injected so
 *  tests never hit the network. */
export type LiveNoteGenerator = (prompt: string, systemPrompt: string) => Promise<string>

/** Local-only persistence for notes + their session anchor. Injected so tests
 *  never touch Electron IPC / SQLite. */
export interface LiveNoteStorage {
  createSession(session: { id: string; startedAt: number; createdAt: number }): Promise<void>
  createNote(note: LiveNote): Promise<void>
  updateNote(id: string, text: string, updatedAt: number): Promise<void>
  deleteNote(id: string): Promise<void>
  listNotes(sessionId: string): Promise<LiveNote[]>
}

/** Minimal read/subscribe surface of the live transcript store (injectable). */
export interface LiveTranscriptSource {
  getSegments(): TranscriptLine[]
  getStatus(): LiveStatus
  isSaved(): boolean
  subscribe(cb: () => void): () => void
}

const defaultGenerator: LiveNoteGenerator = (prompt, systemPrompt) =>
  geminiGenerate({
    model: NOTE_MODEL,
    parts: [{ text: prompt }],
    systemPrompt,
    thinkingBudget: 0
  })

const defaultStorage: LiveNoteStorage = {
  createSession: (s) => window.omi.createTranscriptionSession(s),
  createNote: (n) => window.omi.createLiveNote(n),
  updateNote: (id, text, updatedAt) => window.omi.updateLiveNote(id, text, updatedAt),
  deleteNote: (id) => window.omi.deleteLiveNote(id),
  listNotes: (sessionId) => window.omi.listLiveNotes(sessionId)
}

function newId(): string {
  return crypto.randomUUID()
}

// Map the transcript store's lines to accumulator segments, giving id-less lines a
// stable positional key (the store only appends/upserts, never reorders, so index
// is stable for a given line).
function toAccumulatorSegments(segments: TranscriptLine[]): { id: string; text: string }[] {
  return segments.map((s, i) => ({ id: s.id ?? `#${i}`, text: s.text }))
}

export class LiveNotesMonitor {
  private notes: LiveNote[] = []
  private generating = false
  private aiEnabled = true
  private currentSessionId: string | null = null
  private accumulator = new LiveNotesAccumulator()
  private readonly subscribers = new Set<() => void>()
  private unsubscribeTranscript: (() => void) | null = null
  private startCount = 0

  constructor(
    private readonly generator: LiveNoteGenerator = defaultGenerator,
    private readonly storage: LiveNoteStorage = defaultStorage,
    private readonly live: LiveTranscriptSource = liveConversation
  ) {}

  // --- Public read surface (the panel subscribes like it does to liveConversation) ---

  getNotes(): LiveNote[] {
    return this.notes
  }
  isGenerating(): boolean {
    return this.generating
  }
  isAiEnabled(): boolean {
    return this.aiEnabled
  }
  subscribe(cb: () => void): () => void {
    this.subscribers.add(cb)
    return () => {
      this.subscribers.delete(cb)
    }
  }

  private notify(): void {
    this.subscribers.forEach((cb) => cb())
  }

  // --- Lifecycle (mounted once by LiveNotesHost; refcounted for StrictMode) ---

  /** Begin observing the transcript. Refcounted so a StrictMode double-mount
   *  subscribes exactly once (a second subscription would double-generate). */
  start(): () => void {
    this.startCount++
    if (this.startCount === 1) {
      this.unsubscribeTranscript = this.live.subscribe(() => this.onTranscriptChange())
      // Apply the current state immediately (a session may already be live).
      this.onTranscriptChange()
    }
    return () => this.stop()
  }

  private stop(): void {
    this.startCount = Math.max(0, this.startCount - 1)
    if (this.startCount === 0) {
      this.unsubscribeTranscript?.()
      this.unsubscribeTranscript = null
    }
  }

  private onTranscriptChange(): void {
    const segments = this.live.getSegments()
    const saved = this.live.isSaved()

    // End: the conversation finalized (saved — keep its notes visible) or the
    // session was torn down (segments cleared — clear the list too).
    if (this.currentSessionId !== null && (saved || segments.length === 0)) {
      this.endSession(segments.length === 0 && !saved)
    }

    // Start: a new transcript is flowing with no active session.
    if (this.currentSessionId === null && segments.length > 0 && !saved) {
      this.startSession()
    }

    if (this.currentSessionId === null || !this.aiEnabled) return

    const request = this.accumulator.handleSegmentsUpdate(
      toAccumulatorSegments(segments),
      this.generating
    )
    if (request) void this.generateNote(request)
  }

  private startSession(): void {
    const id = newId()
    const now = Date.now()
    this.currentSessionId = id
    this.notes = []
    this.generating = false
    this.accumulator.reset()
    // Persist the anchor (best-effort — a failed insert must not stop generation;
    // FKs are off so a note with a missing session row still inserts).
    void this.storage.createSession({ id, startedAt: now, createdAt: now }).catch((e) => {
      console.warn('[live-notes] createSession failed:', (e as Error).message)
    })
    this.notify()
  }

  /** End the current session. Keep the notes visible when finalized (saved); clear
   *  them on a hard teardown so the next session starts blank. */
  private endSession(clearNotes: boolean): void {
    this.currentSessionId = null
    this.generating = false
    this.accumulator.reset()
    if (clearNotes && this.notes.length) {
      this.notes = []
      this.notify()
    }
  }

  // --- Manual notes (from the panel) ---

  /** Add a user-typed note. Mints a session if none is active yet (a note typed
   *  while connecting must not be lost — a safe improvement over Mac, which drops
   *  it). Feeds the accumulator context so the AI won't repeat it. */
  async addManualNote(text: string): Promise<void> {
    const trimmed = text.trim()
    if (!trimmed) return
    const sessionId = this.ensureSession()
    const now = Date.now()
    const note: LiveNote = {
      id: newId(),
      sessionId,
      text: trimmed,
      isAi: false,
      segStart: this.accumulator.segmentOrder,
      segEnd: null,
      createdAt: now,
      updatedAt: now
    }
    try {
      await this.storage.createNote(note)
    } catch (e) {
      console.warn('[live-notes] add manual note failed:', (e as Error).message)
      return
    }
    if (this.currentSessionId !== sessionId) return // session changed mid-await
    this.notes = [...this.notes, note]
    this.accumulator.appendExistingNote(trimmed)
    this.notify()
  }

  /** Edit a note's text (explicit user action). Reseeds accumulator context. */
  async updateNote(id: string, text: string): Promise<void> {
    const trimmed = text.trim()
    if (!trimmed) return
    const now = Date.now()
    try {
      await this.storage.updateNote(id, trimmed, now)
    } catch (e) {
      console.warn('[live-notes] update note failed:', (e as Error).message)
      return
    }
    this.notes = this.notes.map((n) => (n.id === id ? { ...n, text: trimmed, updatedAt: now } : n))
    this.accumulator.seedExistingNotes(this.notes.map((n) => n.text))
    this.notify()
  }

  /** Delete a note (explicit user action). Reseeds accumulator context. */
  async deleteNote(id: string): Promise<void> {
    try {
      await this.storage.deleteNote(id)
    } catch (e) {
      console.warn('[live-notes] delete note failed:', (e as Error).message)
      return
    }
    this.notes = this.notes.filter((n) => n.id !== id)
    this.accumulator.seedExistingNotes(this.notes.map((n) => n.text))
    this.notify()
  }

  setAiEnabled(enabled: boolean): void {
    if (this.aiEnabled === enabled) return
    this.aiEnabled = enabled
    this.notify()
  }

  private ensureSession(): string {
    if (this.currentSessionId) return this.currentSessionId
    this.startSession()
    return this.currentSessionId as unknown as string
  }

  // --- AI generation (single-flight; the accumulator already gated the trigger) ---

  private async generateNote(request: LiveNotesGenerationRequest): Promise<void> {
    const sessionId = this.currentSessionId
    if (sessionId === null || !this.aiEnabled || this.generating) return
    this.generating = true
    this.notify()

    const prompt = `Transcript segment:\n${request.recentText}\n\n${request.existingNotesText}\n\n${NOTE_GENERATION_PROMPT}`

    try {
      const raw = await this.generator(prompt, SYSTEM_PROMPT)
      // Clean up (Mac: trim, strip straight quotes/apostrophes).
      const noteText = raw.trim().replace(/["']/g, '')
      if (!noteText) {
        this.finishGeneration(sessionId)
        return
      }
      if (this.currentSessionId !== sessionId) {
        this.finishGeneration(sessionId)
        return
      }
      const now = Date.now()
      const note: LiveNote = {
        id: newId(),
        sessionId,
        text: noteText,
        isAi: true,
        segStart: request.segmentStartOrder,
        segEnd: request.segmentEndOrder,
        createdAt: now,
        updatedAt: now
      }
      await this.storage.createNote(note)
      if (this.currentSessionId !== sessionId) {
        this.finishGeneration(sessionId)
        return
      }
      this.notes = [...this.notes, note]
      this.accumulator.markGenerationSucceeded(noteText)
      this.finishGeneration(sessionId)
    } catch (e) {
      // Generation (or its persist) failed. Degrade gracefully: the transcript is
      // untouched and keeps working, we simply skip this note and the accumulator
      // will retry on the next threshold. This is a provider/fail-open branch, so
      // record it as a fallback (AGENTS.md → Fallback / resilience telemetry). The
      // Windows app has no recordFallback wrapper; the established idiom is a
      // trackEvent('fallback_triggered', …) with the same fields.
      trackEvent('fallback_triggered', {
        component: 'live_notes',
        from: 'gemini',
        to: 'none',
        reason: 'generation_failed',
        outcome: 'degraded'
      })
      console.warn('[live-notes] generation failed:', (e as Error).message)
      this.finishGeneration(sessionId)
    }
  }

  private finishGeneration(sessionId: string): void {
    // Only clear the flag if we still own this session (a teardown may have
    // reset it). Always notify so a spinner stops.
    if (this.currentSessionId === sessionId || this.currentSessionId === null) {
      this.generating = false
    }
    this.notify()
  }
}

// The app-wide singleton the host mounts and the panel subscribes to.
export const liveNotesMonitor = new LiveNotesMonitor()
